Add an inspection that reports properties being set to their default value

matthew16550 opened the pull request.

matthew16550 committed:
Add an inspection that reports properties being set to their default value

matthew16550 commented:
These providers had errors when I ran the schema extractor but I ignored them as out of scope:
* terraform-provider-azure-classic
* terraform-provider-google-beta
* terraform-provider-oci
* terraform-provider-scaffolding

And there were 4 new providers which I also ignored:
* bigip
* netlify
* rightscale
* tencentcloud

matthew16550 commented on schemas-extractor/build-all.sh:
```diff
@@ -52,10 +56,10 @@ for p in $(cat "$CUR/providers.list.full"); do
-  sed -i "s/__NAME__/${p:19}/g" generate-schema/generate-schema.go
-  sed -i "s/__REVISION__/$revision/g" generate-schema/generate-schema.go
-  sed -i "s~__OUT__~$out~g" generate-schema/generate-schema.go
+  sed -i '' "s/__FULL_NAME__/$p/g" generate-schema/generate-schema.go
```
`sed` on my Mac needs this

matthew16550 commented on schemas-extractor/template/generate-schema.go:
```diff
@@ -127,8 +126,11 @@ func export(v *schema.Schema) SchemaDefinition {
 	}
 
 	// TODO: Find better solution
-	if defValue, err := v.DefaultValue(); err == nil && defValue != nil && !reflect.DeepEqual(defValue, v.Default) {
```
I couldn't see what `DeepEqual` was achieving here, perhaps something important that I misunderstood?

VladRassokhin replied:
Can't remember why I've added it here. Does removing changes anything?

matthew16550 replied:
Using `!reflect.DeepEqual(...` there are 265 defaults generated.  Removing it makes 3134 defaults generated.    The diff is too big to compare everything but I think it is correctly finding all default values when removed.

VladRassokhin commented on src/kotlin/org/intellij/plugins/hcl/terraform/config/inspection/TFDefaultPropertyInspection.kt:
```diff
@@ -0,0 +1,63 @@
+            }
+        }
+
+        private fun compare(value: HCLValue, defaultValue: Any): Boolean {
```
It's better to name it `isEquals`

VladRassokhin commented on src/kotlin/org/intellij/plugins/hcl/terraform/config/model/TypeModelLoader.kt:
```diff
@@ -347,7 +348,24 @@ class TypeModelLoader(val external: Map<String, TypeModelProvider.Additional>) {
+    val defaultObj = value.obj("Default")
+    val defaultValue = if (defaultObj == null || defaultObj.isEmpty()) {
+      null
+    }
```
Please reformat code

VladRassokhin commented on src/kotlin/org/intellij/plugins/hcl/terraform/config/model/TypeModelLoader.kt:
```diff
@@ -347,7 +348,24 @@ class TypeModelLoader(val external: Map<String, TypeModelProvider.Additional>) {
     val deprecated = value.string("Deprecated")
-    val has_default: Boolean = value.obj("Default")?.isNotEmpty() ?: false
+    val defaultObj = value.obj("Default")
+    val defaultValue = if (defaultObj == null || defaultObj.isEmpty()) {
```
Seems `if` could be simplified to just `defaultObj?.let { ... }` since there `null -> null` block in `when` which will sort out `isEmpty` case

VladRassokhin commented on src/kotlin/org/intellij/plugins/hcl/terraform/config/model/TypeModelLoader.kt:
```diff
@@ -347,7 +348,24 @@ class TypeModelLoader(val external: Map<String, TypeModelProvider.Additional>) {
+        "Float", "TypeFloat" -> defaultValueString ?: "0"
+        "String", "TypeString" -> defaultValueString ?: ""
+        else -> {
+          LOG.warn("Unhandled default type $typeString for $fqn")
```
`logErrorAndFailInInternalMode` could be used where to detect unsupported types in tests

VladRassokhin commented:
Thank you for contribution.

First, could you please sign [JetBrains CLA](http://www.jetbrains.com/agreements/cla/)

About PR itself I see there're some useless parts of shema being generated.
```
        "Default": {
          "Type": "string"
        }
```
Apart from that I'd like to have separate commit for code and shema update (or even no schema update since it's hard to navigate through PR diff and I'd anyway regenerate them just before releasing next version)

matthew16550 committed:
Also generate Default.Value when Value is the empty string.

matthew16550 committed:
Addressing PR comments:
- Rename compare() -> isEquals()
- Use logErrorAndFailInInternalMode() to report error

Also changed "application" to private class property in TypeModelLoader to avoid passing it around lots of methods.

matthew16550 committed:
Remove schema files from PR

matthew16550 commented on schemas-extractor/template/generate-schema.go:
```diff
@@ -190,13 +194,39 @@ func DoGenerate(provider *schema.Provider, providerName string, outputFilePath s
+}
+
+func (e *SchemaElement) MarshalJSON() ([]byte, error) {
+	if e.Value == nil {
```
Maybe there is a better way?  I don't know much about Go.

VladRassokhin replied:
Neither do I :)

werwolfby replied:
Origin
```
Value string `json:",omitempty"`
```
Should do the trick. You don't need to specify separate Marshaller for this.

And `string` in go is actually `[]byte` which is actually already pointer type. And in most cases it is better to use just `string` instead of `*string`. And compare `string` with empty string (`""`) instead of null. What is literally the same.

I don't understand why original `omitempty` doesn't work.

matthew16550 commented:
I've addressed all the PR comments.

The useless schema like below is because of `json:",omitempty"`,  I've put in a kludge to work around it.

```
"Default": {
  "Type": "string"
}
```

matthew16550 committed:
Merge branch 'master' into inspect-property-defaults
